### PR TITLE
added logging of unused tags

### PR DESCRIPTION
**Pull request by fabpot**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #11511, #11744 |
| License | MIT |
| Doc PR | - |

This is the same as #11744 but with some minor tweaks and some unit tests for the compiler pass.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -1089,6 +1095,19 @@ public function findTags()
...
+     * Returns all tags not queried by findTaggedServiceIds
+     *
+     * @return array An array of tags
+     */
+    public function findUnusedTags()
+    {
+        $tags = array_values(array_diff($this->findTags(), $this->usedTags));
+        $tags = array_unique($tags);
+
+        return $tags;
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: `return array_unique($tags)` (overwriting variables usually not nice)

Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: actually, the array_unique is useless because `findTags` already ensures that

**Comment by Tobion** [comment]
IMO it's enough to add the `UnusedTagsPass` only in debug mode.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ */
+class UnusedTagsPass implements CompilerPassInterface
+{
+    /**
+     * whitelisted tags
+     *
+     * @var array
+     */
+    protected $whitelist = array(
+        "console.command",
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: use single quotes please

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ * @author Florian Pfitzer <pfitzer@wurzel3.de>
+ */
+class UnusedTagsPass implements CompilerPassInterface
+{
+    /**
+     * whitelisted tags
+     *
+     * @var array
+     */
+    protected $whitelist = array(
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: I wouldn't make it protected by default. If we want to offer an extension point, I'd recommend a separate method to add a tag to the whitelist.

Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: Changed to private. I've just added some tests to an existing PR, but that should have been private from the get go.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php (hunk 2 above)
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: indeed

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: yep

**Comment by fabpot** [comment]
Indeed, the pass is now only enabled on debug. That makes sense as the compile log is only dumped in debug mode anyway.

Should be ready now.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 4]
```diff
@@ -1089,6 +1095,16 @@ public function findTags()
         return array_unique($tags);
     }
 
+    /**
+     * Returns all tags not queried by findTaggedServiceIds
+     *
+     * @return array An array of tags
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: lets use `string[]` to be more specific

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 5]
```diff
@@ -91,6 +91,11 @@ class ContainerBuilder extends Container implements TaggedContainerInterface
     private $expressionLanguageProviders = array();
 
     /**
+     * @var array with tag names used by findTaggedServiceIds
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: string[]

**Comment by Tobion** [comment]
:+1: 

Status: Reviewed

**Comment by fabpot** [comment]
Thank you @Marmelatze.

**Pull request merged**